metrics/bigquery.py

```diff
@@ -50,67 +50,159 @@ def validate_metric_name(name):
+
+        # Currently the only modification that is supported is injecting the
+        # timestamp of the most recent influxdb data for a given metric.
+        # (For backfilling)
+        if (self.influx and
+                'measurements' in config and
+                'backfill' in config['measurements'] and
+                config['measurements']['backfill']):
+            # get most recent time
+            measure = config['measurements']['backfill']
+            points = self.influx.query("show field keys from %s limit 1" % measure)
+            points = list(points.get_points())
+
+            field = points[0].get('fieldKey', None) if points else None
+            last_time = None
+            if field:
+                results = self.influx.query(
+                    "select last(%s), time from %s limit 1" % (field, measure)
+                )
+                last_time = next(results.get_points(), {}).get('time', None)
+                if last_time:
+                    # format time properly
+                    last_time = time.strptime(last_time, '%Y-%m-%dT%H:%M:%SZ')
+                    last_time = calendar.timegm(last_time)
+            if not last_time:
+                last_time = int(time.time() - (60*60*24*self.backfill_days))
+            # replace tag with formatted time
+            config['query'] = config['query'].replace('<LAST_DATA_TIME>', str(last_time))
 
 def all_configs(search='**.yaml'):
     """Returns config files in the metrics dir."""
     return glob.glob(os.path.join(
         os.path.dirname(__file__), 'configs', search))
 
-def main(configs, project, bucket_path):
-    """Loads metric config files and runs each metric."""
-    if not project:
-        raise ValueError('project', project)
-    if not bucket_path:
-        raise ValueError('bucket_path', bucket_path)
-
+def make_influx_client():
+    """Make an InfluxDB client from config at path $VELODROME_INFLUXDB_CONFIG"""
     if 'VELODROME_INFLUXDB_CONFIG' in os.environ:
-        influx = influxdb.Pusher.from_config(os.environ['VELODROME_INFLUXDB_CONFIG'])
-    else:
-        influx = None
+        with open(os.environ['VELODROME_INFLUXDB_CONFIG']) as config_file:
+            config = json.load(config_file)
+
+        def check_config(field):
+            if not field in config:
+                raise ValueError('DB client config needs field \'%s\'' % field)
+        check_config('host')
+        check_config('port')
+        check_config('user')
+        check_config('password')
+        return influxdb.InfluxDBClient(
+            host=config['host'],
+            port=config['port'],
+            username=config['user'],
+            password=config['password'],
+            database='metrics'
+        )
+    return None
+
+def ints_to_floats(point):
+    for key, val in point.iteritems():
+        if key == 'time':
+            continue
+        if isinstance(val, int):
+            point[key] = float(val)
```

Comment:
Modifying a dict while iterating through it?! I'm somewhat surprised this works!

Reply:
I'm not changing the keys of the dict (or any sub-dicts), just the values so this is safe because the thing I'm iterating over (the keys) never changes.